File: SQL/Expressions/SQLExpression.vb

' ___________________________________________________
'
'  © Hi-Integrity Systems 2010. All rights reserved.
'  www.hisystems.com.au - Toby Wicks
' ___________________________________________________
'

Option Strict On
Option Explicit On

Namespace SQL

    Public MustInherit Class SQLExpression

        Friend MustOverride Function SQL(ByVal eConnectionType As Database.ConnectionType) As String

        Public Shared Operator +(left As SQLExpression, right As SQLExpression) As SQLArithmeticExpression

            Return New SQLArithmeticExpression(left, ArithmeticOperator.Add, right)

        End Operator

        Public Shared Operator +(left As SQLExpression, right As Decimal) As SQLArithmeticExpression

            Return New SQLArithmeticExpression(left, ArithmeticOperator.Add, New SQLValueExpression(right))

        End Operator

        Public Shared Operator +(left As Decimal, right As SQLExpression) As SQLArithmeticExpression

            Return New SQLArithmeticExpression(New SQLValueExpression(left), ArithmeticOperator.Add, right)

        End Operator

        Public Shared Operator +(left As SQLExpression, right As DateTime) As SQLArithmeticExpression

            Return New SQLArithmeticExpression(left, ArithmeticOperator.Add, New SQLValueExpression(right))

        End Operator

        Public Shared Operator +(left As DateTime, right As SQLExpression) As SQLArithmeticExpression

            Return New SQLArithmeticExpression(New SQLValueExpression(left), ArithmeticOperator.Add, right)

        End Operator

        Public Shared Operator +(left As String, right As SQLExpression) As SQLStringConcatExpression

            Return New SQLStringConcatExpression(New SQLValueExpression(left), right)

        End Operator

        Public Shared Operator +(left As SQLExpression, right As String) As SQLStringConcatExpression

            Return New SQLStringConcatExpression(left, New SQLValueExpression(right))

        End Operator

        Public Shared Operator +(left As SQLStringConcatExpression, right As SQLExpression) As SQLStringConcatExpression

            Return New SQLStringConcatExpression(left, right)

        End Operator

        Public Shared Operator +(left As SQLExpression, right As SQLStringConcatExpression) As SQLStringConcatExpression

            Return New SQLStringConcatExpression(left, right)

        End Operator

        Public Shared Operator &(left As String, right As SQLExpression) As SQLStringConcatExpression

            Return New SQLStringConcatExpression(New SQLValueExpression(left), right)

        End Operator

        Public Shared Operator &(left As SQLExpression, right As String) As SQLStringConcatExpression

            Return New SQLStringConcatExpression(left, New SQLValueExpression(right))

        End Operator

        Public Shared Operator &(left As SQLStringConcatExpression, right As SQLExpression) As SQLStringConcatExpression

            Return New SQLStringConcatExpression(left, right)

        End Operator

        Public Shared Operator &(left As SQLExpression, right As SQLStringConcatExpression) As SQLStringConcatExpression

            Return New SQLStringConcatExpression(left, right)

        End Operator

        Public Shared Operator -(left As SQLExpression, right As SQLExpression) As SQLArithmeticExpression

            Return New SQLArithmeticExpression(left, ArithmeticOperator.Subtract, right)

        End Operator

        Public Shared Operator -(left As SQLExpression, right As Decimal) As SQLArithmeticExpression

            Return New SQLArithmeticExpression(left, ArithmeticOperator.Subtract, New SQLValueExpression(right))

        End Operator

        Public Shared Operator -(left As Decimal, right As SQLExpression) As SQLArithmeticExpression

            Return New SQLArithmeticExpression(New SQLValueExpression(left), ArithmeticOperator.Subtract, right)

        End Operator

        Public Shared Operator -(left As SQLExpression, right As DateTime) As SQLArithmeticExpression

            Return New SQLArithmeticExpression(left, ArithmeticOperator.Subtract, New SQLValueExpression(right))

        End Operator

        Public Shared Operator -(left As DateTime, right As SQLExpression) As SQLArithmeticExpression

            Return New SQLArithmeticExpression(New SQLValueExpression(left), ArithmeticOperator.Subtract, right)

        End Operator

        Public Shared Operator *(left As SQLExpression, right As SQLExpression) As SQLArithmeticExpression

            Return New SQLArithmeticExpression(left, ArithmeticOperator.Multiply, right)

        End Operator

        Public Shared Operator *(left As SQLExpression, right As Decimal) As SQLArithmeticExpression

            Return New SQLArithmeticExpression(left, ArithmeticOperator.Multiply, New SQLValueExpression(right))

        End Operator

        Public Shared Operator *(left As Decimal, right As SQLExpression) As SQLArithmeticExpression

            Return New SQLArithmeticExpression(New SQLValueExpression(left), ArithmeticOperator.Multiply, right)

        End Operator

        Public Shared Operator /(left As SQLExpression, right As SQLExpression) As SQLArithmeticExpression

            Return New SQLArithmeticExpression(left, ArithmeticOperator.Divide, right)

        End Operator

        Public Shared Operator /(left As SQLExpression, right As Decimal) As SQLArithmeticExpression

            Return New SQLArithmeticExpression(left, ArithmeticOperator.Divide, New SQLValueExpression(right))

        End Operator

        Public Shared Operator /(left As Decimal, right As SQLExpression) As SQLArithmeticExpression

            Return New SQLArithmeticExpression(New SQLValueExpression(left), ArithmeticOperator.Divide, right)

        End Operator

        Public Shared Operator Mod(left As SQLExpression, right As SQLExpression) As SQLArithmeticExpression

            Return New SQLArithmeticExpression(left, ArithmeticOperator.Modulus, right)

        End Operator

        Public Shared Operator Mod(left As SQLExpression, right As Decimal) As SQLArithmeticExpression

            Return New SQLArithmeticExpression(left, ArithmeticOperator.Modulus, New SQLValueExpression(right))

        End Operator

        Public Shared Operator Mod(left As Decimal, right As SQLExpression) As SQLArithmeticExpression

            Return New SQLArithmeticExpression(New SQLValueExpression(left), ArithmeticOperator.Modulus, right)

        End Operator

        Public Shared Operator Or(left As SQLExpression, right As SQLExpression) As SQLBitwiseExpression

            Return New SQLBitwiseExpression(left, BitwiseOperator.Or, right)

        End Operator

        Public Shared Operator Or(left As SQLExpression, right As Decimal) As SQLBitwiseExpression

            Return New SQLBitwiseExpression(left, BitwiseOperator.Or, New SQLValueExpression(right))

        End Operator

        Public Shared Operator Or(left As Decimal, right As SQLExpression) As SQLBitwiseExpression

            Return New SQLBitwiseExpression(New SQLValueExpression(left), BitwiseOperator.Or, right)

        End Operator

        Public Shared Operator And(left As SQLExpression, right As SQLExpression) As SQLBitwiseExpression

            Return New SQLBitwiseExpression(left, BitwiseOperator.And, right)

        End Operator

        Public Shared Operator And(left As SQLExpression, right As Decimal) As SQLBitwiseExpression

            Return New SQLBitwiseExpression(left, BitwiseOperator.And, New SQLValueExpression(right))

        End Operator

        Public Shared Operator And(left As Decimal, right As SQLExpression) As SQLBitwiseExpression

            Return New SQLBitwiseExpression(New SQLValueExpression(left), BitwiseOperator.And, right)

        End Operator

    End Class

End Namespace

File: SQL/Expressions/SQLExpression.cs
// _________________________________________________________________________
//
//  © Hi-Integrity Systems 2010. All rights reserved.
//  www.hisystems.com.au - Toby Wicks
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//	    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// _________________________________________________________________________
//

using System.Collections;
using System;
using System.Data;

namespace DatabaseObjects.SQL
{
    public abstract class SQLExpression
    {
        internal abstract string SQL(Serializers.Serializer serializer);

        public static SQLArithmeticExpression operator +(SQLExpression left, SQLExpression right)
        {
            return new SQLArithmeticExpression(left, ArithmeticOperator.Add, right);
        }

        public static SQLArithmeticExpression operator +(SQLExpression left, decimal right)
        {
            return new SQLArithmeticExpression(left, ArithmeticOperator.Add, new SQLValueExpression(right));
        }

        public static SQLArithmeticExpression operator +(decimal left, SQLExpression right)
        {
            return new SQLArithmeticExpression(new SQLValueExpression(left), ArithmeticOperator.Add, right);
        }

        public static SQLArithmeticExpression operator +(SQLExpression left, DateTime right)
        {
            return new SQLArithmeticExpression(left, ArithmeticOperator.Add, new SQLValueExpression(right));
        }

        public static SQLArithmeticExpression operator +(DateTime left, SQLExpression right)
        {
            return new SQLArithmeticExpression(new SQLValueExpression(left), ArithmeticOperator.Add, right);
        }

        public static SQLStringConcatExpression operator +(string left, SQLExpression right)
        {
            return new SQLStringConcatExpression(new SQLValueExpression(left), right);
        }

        public static SQLStringConcatExpression operator +(SQLExpression left, string right)
        {
            return new SQLStringConcatExpression(left, new SQLValueExpression(right));
        }

        public static SQLStringConcatExpression operator +(SQLStringConcatExpression left, SQLExpression right)
        {
            return new SQLStringConcatExpression(left, right);
        }

        public static SQLStringConcatExpression operator +(SQLExpression left, SQLStringConcatExpression right)
        {
            return new SQLStringConcatExpression(left, right);
        }
        
        public static SQLArithmeticExpression operator -(SQLExpression left, SQLExpression right)
        {
            return new SQLArithmeticExpression(left, ArithmeticOperator.Subtract, right);
        }

        public static SQLArithmeticExpression operator -(SQLExpression left, decimal right)
        {
            return new SQLArithmeticExpression(left, ArithmeticOperator.Subtract, new SQLValueExpression(right));
        }

        public static SQLArithmeticExpression operator -(decimal left, SQLExpression right)
        {
            return new SQLArithmeticExpression(new SQLValueExpression(left), ArithmeticOperator.Subtract, right);
        }

        public static SQLArithmeticExpression operator -(SQLExpression left, DateTime right)
        {
            return new SQLArithmeticExpression(left, ArithmeticOperator.Subtract, new SQLValueExpression(right));
        }

        public static SQLArithmeticExpression operator -(DateTime left, SQLExpression right)
        {
            return new SQLArithmeticExpression(new SQLValueExpression(left), ArithmeticOperator.Subtract, right);
        }

        public static SQLArithmeticExpression operator *(SQLExpression left, SQLExpression right)
        {
            return new SQLArithmeticExpression(left, ArithmeticOperator.Multiply, right);
        }

        public static SQLArithmeticExpression operator *(SQLExpression left, decimal right)
        {
            return new SQLArithmeticExpression(left, ArithmeticOperator.Multiply, new SQLValueExpression(right));
        }

        public static SQLArithmeticExpression operator *(decimal left, SQLExpression right)
        {
            return new SQLArithmeticExpression(new SQLValueExpression(left), ArithmeticOperator.Multiply, right);
        }

        public static SQLArithmeticExpression operator /(SQLExpression left, SQLExpression right)
        {
            return new SQLArithmeticExpression(left, ArithmeticOperator.Divide, right);
        }

        public static SQLArithmeticExpression operator /(SQLExpression left, decimal right)
        {
            return new SQLArithmeticExpression(left, ArithmeticOperator.Divide, new SQLValueExpression(right));
        }

        public static SQLArithmeticExpression operator /(decimal left, SQLExpression right)
        {
            return new SQLArithmeticExpression(new SQLValueExpression(left), ArithmeticOperator.Divide, right);
        }

        public static SQLArithmeticExpression operator %(SQLExpression left, SQLExpression right)
        {
            return new SQLArithmeticExpression(left, ArithmeticOperator.Modulus, right);
        }

        public static SQLArithmeticExpression operator %(SQLExpression left, decimal right)
        {
            return new SQLArithmeticExpression(left, ArithmeticOperator.Modulus, new SQLValueExpression(right));
        }

        public static SQLArithmeticExpression operator %(decimal left, SQLExpression right)
        {
            return new SQLArithmeticExpression(new SQLValueExpression(left), ArithmeticOperator.Modulus, right);
        }

        public static SQLBitwiseExpression operator |(SQLExpression left, SQLExpression right)
        {
            return new SQLBitwiseExpression(left, BitwiseOperator.Or, right);
        }

        public static SQLBitwiseExpression operator |(SQLExpression left, decimal right)
        {
            return new SQLBitwiseExpression(left, BitwiseOperator.Or, new SQLValueExpression(right));
        }

        public static SQLBitwiseExpression operator |(decimal left, SQLExpression right)
        {
            return new SQLBitwiseExpression(new SQLValueExpression(left), BitwiseOperator.Or, right);
        }

        public static SQLBitwiseExpression operator &(SQLExpression left, SQLExpression right)
        {
            return new SQLBitwiseExpression(left, BitwiseOperator.And, right);
        }

        public static SQLBitwiseExpression operator &(SQLExpression left, decimal right)
        {
            return new SQLBitwiseExpression(left, BitwiseOperator.And, new SQLValueExpression(right));
        }

        public static SQLBitwiseExpression operator &(decimal left, SQLExpression right)
        {
            return new SQLBitwiseExpression(new SQLValueExpression(left), BitwiseOperator.And, right);
        }
    }
}